### PR TITLE
feat(masking): support PII detection: idCard, phone, email, ipAddress, and bankCard

### PR DESCRIPTION
**Pull request by Monsooooon**
## Summary

- add configurable masking for `idCard`, `phone`, `email`, `ipAddress`, and `bankCard`
- validate candidates with mainland China ID checksum/date rules, mainland phone formats, ASCII email structure, IPv4 segment checks, and issuer/Luhn card checks
- reuse the existing pre-output masking path so JSONL, SLS, HTTP, and OTLP Trace receive the same masked content
- keep numeric scanning shared and bounded, and build multi-match replacements in one pass to reduce CPU and allocations
- document the supported types, replacement markers, and phase-one detection boundaries

## Behavior

- `mask.mode: all` enables the existing secret rules and all five PII types
- `mask.mode: custom` runs only explicitly selected types
- the default remains `mask.mode: none`
- scanning remains limited to the existing content-field whitelist

## Performance

Development-machine microbenchmarks show about `0.74 ms` average / `0.84 ms` P95 for a roughly 75 KiB mixed sample. For a dense 19 KiB sample with 1,250 matches, single-pass replacement reduced average time from `2.01 ms` to `0.71 ms`.

## Validation

- `npx vitest run tests/unit/mask tests/unit/core/config-loader.test.ts tests/unit/core/input-manager.test.ts tests/performance/pii-mask.perf.test.ts`
- `LOG_LEVEL=silent npx vitest run --silent --reporter=basic`
- `npm run typecheck`
- `npm run build`
- `git diff --cached --check`

### CODE REVIEW

File: src/mask/string-masker.ts [hunk 1]
```diff
@@ -248,14 +278,40 @@ function normalizeMaskRanges(
...
-    if (range.start < lastEnd) continue;
-    result.push(range);
-    lastEnd = range.end;
+  for (const candidate of sorted) {
+    const previous = result[result.length - 1];
+    if (!previous || candidate.start >= previous.end) {
+      result.push(candidate);
+      continue;
+    }
+    if (getMaskRangePriority(candidate) > getMaskRangePriority(previous)) {
```
Comment (linrunqi08) [inline src/mask/string-masker.ts:298]: **[High]** `normalizeMaskRanges` 处理重叠 range 时整段替换（`result[len-1] = candidate`），只保留其一，而非两者覆盖区间的并集。当高优先级 range 嵌套在低优先级 range 内部且更短时，低优先级 range 未被覆盖的头/尾字节被留在明文中。

**影响:** 已被检出的 PII 明文泄漏。实测（真实检测器代码）：`"user.13800138000@example.com"` → `"user.[PHONE_MASKED]@example.com"`（本地名 `user.` 与完整域名 `@example.com` 泄漏）；`"11010519491231002X@example.com"` → `"[IDCARD_MASKED]@example.com"`（域名泄漏）。密钥规则（优先级 100）嵌入 PII 时同样走此分支。`mode:all` 下稳定触发。这是本 PR 新增优先级替换分支引入的新缺陷。

**建议:** 重叠合并必须保证并集全覆盖——重叠时保留区间并集（`start=min,end=max`），优先级仅决定重叠段采用哪个 replacement；或对超出部分追加裁剪后的子区间。并补「短高优先级嵌入长低优先级」单测，断言输出无原文残留。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/mask/pii-detectors.ts [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+    if (index > rawStart && isNumericCandidateStart(value.charCodeAt(index - 1))) {
+      ends.push(index);
+    }
+    if (index + 1 < end && isNumericCandidateStart(value.charCodeAt(index + 1))) {
+      starts.push(index + 1);
+    }
+  }
+  ends.push(end);
+
+  for (const start of starts) {
```
Comment (linrunqi08) [inline src/mask/pii-detectors.ts]: **[High]** `collectFormattedNumericCandidate` 对每个 ≤128 字符数字候选窗口，按空格切出 `starts`/`ends` 后做 `starts × ends` 笛卡尔积双重循环，每个子串 `slice` + 手机正则 + `replace` + Luhn 校验，最坏 O(k²)。

**影响:** 可控输入触发的 CPU 放大/DoS。masking 位于遥测采集同步路径（prompt/completion/tool 参数与结果均过 `maskString`，内容上游可控）。实测（真实代码）：64KiB 对抗输入耗时 **1089ms**，同规模良性文本仅 **7.7ms**（~140× 放大），单字段即可阻塞事件循环 1s+；空格分隔的数字表格（矩阵/CSV 落日志）也天然命中。

**建议:** 用线性滑窗替代笛卡尔积——只对「相邻 token 合并且总长落在 phone/bankCard 合法长度区间」的连续片段校验；或对单窗口 `starts.length/ends.length` 设硬上限（>8 只验整段）；并将 `MAX_NUMERIC_SEQUENCE_LENGTH` 收敛到实际上限（≤32）。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/mask/pii-detectors.ts [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+
+  const topLevelDomain = labels[labels.length - 1];
+  if (topLevelDomain.length < 2) return false;
+  for (let index = 0; index < topLevelDomain.length; index += 1) {
+    if (!isAsciiLetter(topLevelDomain.charCodeAt(index))) return false;
+  }
+  return true;
+}
+
+function collectIpAddressRanges(value: string, ranges: MaskRange[]): void {
```
Comment (linrunqi08) [inline src/mask/pii-detectors.ts:486]: **[Medium]** `collectIpAddressRanges` 把任意 `d.d.d.d`（逐段 0-255）判为 IP，含版本号形状 `1.2.3.4`、四段点分标识、全部内网/回环 IP。

**影响:** `mode:all` 下会掩掉日志内几乎所有 IPv4 与版本号，破坏可观测/排障定位数据。docs 已记为 phase-one 边界，故定 Medium，但运行期破坏面较大。

**建议:** 默认模式不启用 ipAddress，或对紧邻 `v`/`version`/`ver` 关键字的四段数字豁免；并在文档显著位置提示对可观测数据的影响。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/mask/pii-detectors.ts [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+      !isAsciiDigit(candidate.charCodeAt(index - 1)) ||
+      !isAsciiDigit(candidate.charCodeAt(index + 1))
+    ) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function matchesCardIssuer(digits: string): boolean {
```
Comment (linrunqi08) [inline src/mask/pii-detectors.ts:359]: **[Medium]** `matchesCardIssuer`/`isValidBankCard` 银联前缀 `62`、Visa `4` 覆盖 16-19 位极广区间 + Luhn（1/10 命中）；`isValidPhone` 对 `1[3-9]` 开头 11 位一律判手机。

**影响:** `mode:all` 下 16-19 位业务/订单 ID（`62`/`4` 开头且偶过 Luhn）或 11 位数值 ID 会被误掩，数值型业务 ID 密集的日志存在关联字段丢失。属检测器固有精度取舍。

**建议:** 文档明确误报边界；如条件允许，对紧邻业务关键字（`orderId`/`traceId` 等）的长数字降权。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/core/config-loader.ts [hunk 5]
```diff
@@ -324,6 +324,11 @@ const SUPPORTED_MASK_TYPES: readonly MaskType[] = [
   'apiKey',
   'privateKey',
   'databaseUrl',
+  'idCard',
```
Comment (linrunqi08) [inline src/core/config-loader.ts]: **[Medium]** 支持类型清单在 3 处手工并列维护：本文件 `SUPPORTED_MASK_TYPES`（config 校验）、`src/mask/rule-loader.ts:22`（展开/过滤）、`src/types/index.ts:36`（`MaskType` 联合类型），缺单一事实源。

**影响:** 后续新增类型漏改 rule-loader 的 `SUPPORTED_MASK_TYPES` 时，`mode:all` 不启用、custom 被静默过滤，无编译错误、无测试报警——静默失效。

**建议:** 以 `MaskType`/`PiiMaskType` 为唯一事实源导出单一 `SUPPORTED_MASK_TYPES`，或加「联合类型成员数 === 清单长度」断言防漂移。

---
*Generated by LoongSuite-Pilot Code Review Agent*

**Comment by linrunqi08** [comment]
## 🔍 Code Review Summary

| Severity | Count |
|----------|-------|
| Critical | 0 |
| High     | 2 |
| Medium   | 3 |
| Low      | 5 |

### Lifecycle Verdict

| Check | Result |
|-------|--------|
| 资源释放 | PASS |
| 死锁/卡死风险 | PASS |
| 状态恢复正确性 | PASS |

> 注：Lifecycle 全 PASS，但 High-2 的 CPU 放大会长时间同步占用事件循环（非死锁，属可用性风险，已计入 Findings）。

### Merge Gate（合入门禁）

⛔ **BLOCK**（阻断级别：`Critical` / `High` / `Medium`，`Low` 不阻断；Lifecycle 任一 FAIL 也阻断）

本轮未解决的阻断项：
- **High** — `src/mask/string-masker.ts:296` `normalizeMaskRanges` 重叠整段替换导致已检出 PII 明文头/尾泄漏
- **High** — `src/mask/pii-detectors.ts:193-219` 数字候选 `starts×ends` 笛卡尔积 O(k²)，可控输入 CPU 放大 ~140×（同步路径 DoS）
- **Medium** — `src/mask/pii-detectors.ts:434` IPv4 over-masking，掩掉版本号/内网 IP 破坏可观测数据
- **Medium** — `src/mask/pii-detectors.ts:279-320` 长数字业务 ID（银行卡/手机形状）误掩
- **Medium** — 支持类型清单 3 处漂移（`config-loader.ts:327` / `rule-loader.ts:22` / `types/index.ts:36`），漏改静默失效

### 总体结论

本 PR 引入 5 类 PII 检测，身份证校验严谨、单遍替换思路正确、向后兼容良好。但存在 **2 个 High**（明文泄漏 + 同步路径 CPU 放大 DoS，均已用真实 bundled 代码复现）与 **3 个 Medium**（over-masking、误掩、类型清单漂移），构成合入阻断。核心风险是「已检出却未完全掩掉」的 PII 泄漏与遥测同步路径的可用性放大——建议优先修复这两项 High 并补对抗性单测/性能用例后再合入。

### Highlights（正向实践）

- 数字类检测共享单次扫描 + `applyMaskRanges` 单遍构建输出，避免逐匹配整串拷贝。
- 身份证校验严谨（省份码 + 日期 + 顺序码非 000 + GB11643 校验位，含末位小写 `x`）。
- `MASKED_TOKEN_PATTERN` 扩展保证已脱敏值幂等；`plan | rules[]` 双签名向后兼容。
- 检测边界在中英文档明确记录（phase-one boundaries），配正/反例与性能用例，41 项 mask 单测通过。

---
*评审报告详见: `code-review/pr-199/final-report.md`*
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/mask/pii-detectors.ts [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+}
+
+function isNumericCandidateStart(code: number): boolean {
+  return isAsciiDigit(code) || code === 43 || code === 40;
+}
+
+function isNumericCandidateChar(code: number): boolean {
+  return isAsciiDigit(code) || code === 32 || code === 43 || code === 45 || code === 40 || code === 41;
+}
+
```
Comment (ralf0131) [inline src/mask/pii-detectors.ts:562]: **[Warning]** `getCurrentDateNumber()` creates a new `Date()` object on every `collectIdCardCandidates` call. In a telemetry pipeline processing many entries per second, this adds up. Consider caching the date number at the `MaskPlan` level or refreshing it once per `maskString` invocation (the date won't change within a single masking operation).

File: src/mask/string-masker.ts [hunk 7]
```diff
@@ -248,14 +278,40 @@ function normalizeMaskRanges(
...
+      result[result.length - 1] = candidate;
+    }
   }
   return result;
```
Comment (ralf0131) [inline src/mask/string-masker.ts:311]: **[Info]** The single-pass chunk-based replacement in `applyMaskRanges` is a significant improvement over the previous iterative concatenation. Good optimization for dense-match scenarios.

File: src/mask/pii-detectors.ts [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+  const topLevelDomain = labels[labels.length - 1];
+  if (topLevelDomain.length < 2) return false;
+  for (let index = 0; index < topLevelDomain.length; index += 1) {
+    if (!isAsciiLetter(topLevelDomain.charCodeAt(index))) return false;
+  }
+  return true;
+}
+
+function collectIpAddressRanges(value: string, ranges: MaskRange[]): void {
+  let index = 0;
```
Comment (ralf0131) [inline src/mask/pii-detectors.ts:487]: **[Info]** The early-return optimization in `collectFormattedNumericCandidate` (trying the full range first before falling back to the cartesian product) is a good mitigation. However, as @linrunqi08 noted, adversarial input with many space-separated numeric segments can still trigger the O(k²) path. Consider adding a hard cap on `starts.length * ends.length` (e.g., 1000) and skipping the cartesian product if exceeded.

File: src/mask/rule-loader.ts [hunk 9]
```diff
@@ -35,10 +57,28 @@ export function loadSensitiveRules(): CompiledMaskRule[] {
```
Comment (ralf0131) [inline src/mask/rule-loader.ts:42]: **[Info]** The `MaskPlan` abstraction cleanly separates secret rules from PII types, making the pipeline more maintainable. The `resolveMaskPlan` backward-compatibility wrapper (accepting both `MaskPlan` and `CompiledMaskRule[]`) is a good migration strategy.

**Review by ralf0131 — COMMENTED** [review]
﻿### Summary
This PR adds PII detection (idCard, phone, email, ipAddress, bankCard) to Pilot's masking pipeline with a clean `MaskPlan` abstraction separating secret rules from PII types. The architecture is well-structured, test coverage is comprehensive (including performance benchmarks), and documentation clearly documents phase-one boundaries.

@linrunqi08 has already provided thorough inline comments on the critical issues (overlap handling bug, cartesian product DoS risk, aggressive IP detection, false positive rates, and type-list maintenance). I agree with all of those findings.

**Additional notes:**
- The `getCurrentDateNumber()` call per mask operation could be cached at the plan level for better performance in high-throughput scenarios
- The early-return optimization before the cartesian product is good, but consider adding a hard cap on iterations as a safety net
- The single-pass chunk-based replacement is a significant performance improvement

**Verdict:** The high-severity issues flagged by @linrunqi08 should be addressed before merge. Once those are resolved, this PR will be a solid addition to the masking pipeline.

---
*Automated review by github-manager-bot*

File: src/mask/pii-detectors.ts [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+function trimNumericCandidateEnd(value: string, start: number, rawEnd: number): number {
+  let end = rawEnd;
+  while (end > start && isTrimmableNumericSuffix(value.charCodeAt(end - 1))) {
+    end -= 1;
+  }
+  return end;
+}
+
+function isValidPhone(candidate: string): boolean {
+  if (candidate.length < 10 || candidate.length > MAX_PHONE_CANDIDATE_LENGTH) {
```
Comment (ralf0131) [inline src/mask/pii-detectors.ts:312]: **[Warning]** Landline phone detection (`isValidPhone`) accepts formats like `010-12345678` and `(010)12345678`. The separator-based matching in `hasValidSimpleSeparators` could potentially match non-phone dash-separated numeric patterns (e.g., product codes `010-1234567`). Consider adding a minimum digit-count heuristic for the segments, or documenting the known false-positive boundary. Tests cover the happy path well, but edge cases with ambiguous separators might warrant a follow-up.

File: src/mask/pii-detectors.ts [hunk 11]
```diff
@@ -0,0 +1,607 @@
...
+      if (digit > 9) digit -= 9;
+    }
+    sum += digit;
+    doubleDigit = !doubleDigit;
+  }
+  return sum % 10 === 0;
+}
+
+function collectEmailRanges(value: string, ranges: MaskRange[]): void {
+  let atIndex = value.indexOf('@');
```
Comment (ralf0131) [inline src/mask/pii-detectors.ts:390]: **[Info]** The email boundary detection in `collectEmailRanges` uses multiple boundary flags (`stoppedByLocalLimit`, `stoppedByTotalLimit`, `touchesAnotherAt`, `trailingPunctuationDot`, `touchesEmailChar`) to disambiguate email addresses embedded in text. The interaction between these conditions is intricate — well done on the thorough test coverage. A brief code comment explaining the priority/ordering of these checks would help future maintainers.

File: src/mask/string-masker.ts [hunk 12]
```diff
@@ -8,33 +9,59 @@ import {
...
+  }
+  if (plan.piiTypes.size > 0) {
+    ranges.push(...collectPiiRanges(value, plan.piiTypes));
+  }
 
   return applyMaskRanges(value, ranges);
 }
 
+function resolveMaskPlan(planOrRules: MaskPlan | readonly CompiledMaskRule[]): MaskPlan {
+  if (Array.isArray(planOrRules)) {
```
Comment (ralf0131) [inline src/mask/string-masker.ts:56]: **[Info]** When PII ranges overlap with secret-rule ranges, the `applyMaskRanges` function handles deduplication via the normalized range merge. It would be helpful to document the effective priority ordering (e.g., `idCard > phone > bankCard > email > ipAddress` based on `collectPiiRanges` ordering) as a code comment, so future contributors understand which detector wins on ambiguous overlaps.

**Review by ralf0131 — APPROVED** [review]
### Summary
This PR adds comprehensive PII detection and masking for 5 content types (idCard, phone, email, ipAddress, bankCard) with thorough validation logic, single-pass replacement for performance, and clean integration into the existing masking pipeline. The code is well-structured, the test coverage is excellent (unit + integration + performance), and the documentation is complete.

**Key strengths:**
- Solid validation: Chinese ID checksum/date rules, Luhn for bank cards, proper email domain label checks
- Performance-conscious: single-pass replacement, bounded scanning, large-string handling
- Clean architecture: `MaskPlan` abstraction separates manifest rules from PII detectors
- Comprehensive tests: 253-line pii-detectors test, edge cases, perf benchmarks

LGTM — approving with minor informational notes above.

---
*Automated review by "github-manager-bot"*

File: src/mask/string-masker.ts (hunk 1 above)
Comment (linrunqi08) [reply src/mask/string-masker.ts:298]: **[fixed]** 重叠合并已改为并集覆盖,泄漏消除。

**证据:** `normalizeMaskRanges`(fix commit d65717a7)高优先级重叠时保留 `start: previous.start, end: max(previous.end, candidate.end)`,低优先级重叠时扩展 `end`;新增单测 `keeps the full union covered across chained overlapping ranges`。真实检测器代码复现:`user.13800138000@example.com` → `[PHONE_MASKED]`、`11010519491231002X@example.com` → `[IDCARD_MASKED]`,无头/尾明文残留。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/mask/pii-detectors.ts (hunk 2 above)
Comment (linrunqi08) [reply src/mask/pii-detectors.ts]: **[fixed]** 笛卡尔积双重循环已改为线性有界滑窗,DoS 放大消除。

**证据:** `collectFormattedNumericCandidate`(d65717a7)按 `MAX_PHONE_CANDIDATE_LENGTH=18`/`MAX_BANK_CARD_CANDIDATE_LENGTH=37` 窗口 + 分隔符步进单遍扫描,并新增首字符预筛(`isPotentialPhoneStart`/`isPotentialBankCardStart`)。实测 64KiB 对抗输入 1089ms → 63.7ms(线性有界);`tests/performance/pii-mask.perf.test.ts` fragmented-numeric 用例 6.7ms 通过,新增 `string-masker.performance.test.ts`。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/mask/pii-detectors.ts (hunk 3 above)
Comment (linrunqi08) [reply src/mask/pii-detectors.ts:486]: **[not-fixed / 已文档化取舍]** 代码行为未变(仍将 `d.d.d.d` 判为 IPv4),但作者已在 `docs/masking.md` 与 `docs/zh-CN/masking.md` 明确记录该 phase-one 边界,并给出缓解路径(数值/版本号密集场景用 `custom` 模式按需开启)。作为 phase-one 精度取舍可接受,**不阻断合入**;建议后续对紧邻 `v`/`version` 的四段数字豁免。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/mask/pii-detectors.ts (hunk 4 above)
Comment (linrunqi08) [reply src/mask/pii-detectors.ts:359]: **[partially-fixed / 已文档化取舍]** 首字符预筛 + 精确位数门控(phone 按起始符校验位数、bankCard 限 15-19 位并需过发卡行前缀+Luhn)已收窄误报面;剩余同形数值 ID 误掩风险已在中英文档明确并给出 `custom` 模式缓解指引。作为精度取舍可接受,**不阻断合入**。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: src/core/config-loader.ts (hunk 5 above)
Comment (linrunqi08) [reply src/core/config-loader.ts]: **[fixed]** 类型清单已收敛为单一事实源。

**证据:** `src/types/index.ts` 以 `PII_MASK_TYPES` / `SUPPORTED_MASK_TYPES`(`as const` 数组)派生 `MaskType`/`PiiMaskType`;`config-loader.ts` 与 `rule-loader.ts` 均改为 import 该常量,`MANIFEST_MASK_TYPES` 由 `SUPPORTED_MASK_TYPES.filter` 计算,3 处手工并列已消除。

---
*Generated by LoongSuite-Pilot Code Review Agent*

**Comment by linrunqi08** [comment]
## 🔍 Code Review Summary — Round 2（增量复审）

- HEAD: `d65717a7`（round-1: `bde62c38`，fix commit `d65717a7 fix(masking): harden overlap and numeric scanning`）

### Round-1 阻断项复核

| # | Sev | 位置 | 状态 | 证据 |
|---|-----|------|------|------|
| 1 | High | string-masker.ts:296 | ✅ fixed | 重叠合并改为并集覆盖;真实代码复现 `user.13800138000@example.com`→`[PHONE_MASKED]`、`11010519491231002X@example.com`→`[IDCARD_MASKED]`,无残留;新增全并集覆盖单测 |
| 2 | High | pii-detectors.ts:193-219 | ✅ fixed | 笛卡尔积→线性有界滑窗+首字符预筛;64KiB 对抗输入 1089ms→63.7ms;perf 用例 6.7ms 通过 |
| 3 | Medium | pii-detectors.ts:434 (IPv4) | 📝 文档化取舍 | 行为未变,已在中英文档记录并给出 `custom` 模式缓解;非安全缺陷,不阻断 |
| 4 | Medium | pii-detectors.ts:279-320 (数值 FP) | 📝 partially-fixed | 位数/前缀门控收窄误报面 + 文档化缓解;不阻断 |
| 5 | Medium | 类型清单 3 处漂移 | ✅ fixed | `types/index.ts` 单一事实源(`as const` 派生),config-loader/rule-loader 均引用 |

### Lifecycle Verdict

| Check | Result |
|-------|--------|
| 资源释放 | PASS |
| 死锁/卡死风险 | PASS |
| 状态恢复正确性 | PASS |

### Merge Gate（合入门禁）

✅ **APPROVE-READY**（本评审 findings 视角）

- 2 个 High(明文泄漏 + 同步路径 CPU 放大)与类型漂移 Medium 已修复并经真实代码/测试验证。
- IPv4 over-masking 与数值业务 ID 误报为 phase-one 精度取舍,作者已文档化并提供 `custom` 模式缓解,不构成阻断。
- 全量 57 项 mask 测试通过(含新增并集覆盖 + 性能用例)。

> 说明:PR 上另有其他评审者的评论线程,不在本 agent 评审范围内,合入前请一并确认。

### 建议(不阻断)

- 后续可对紧邻 `v`/`version` 的四段数字豁免 IPv4 掩码;`validateBaseRule` 目前仍用宽集合 `SUPPORTED_MASK_TYPE_SET` 校验 manifest 规则,可收敛为 `MANIFEST_MASK_TYPES`。

---
*评审报告详见: `code-review/pr-199/final-report.md`*
*Generated by LoongSuite-Pilot Code Review Agent*

**Review by linrunqi08 — APPROVED** [review]
